This github repo has been setup to help foster a more sustainable collaborative
environment in which the workshop content can be developed and refined. To
facilitate this we are:

* Porting all the documentation over to LaTeX - easy to version control and
  easier to collaborate on content. Easy to get consistent styling and can
  (will) be able to parse out code (listings) from within the LaTeX docs
  for easy automated testing of the code.
* Developing the Bioinformatics Training Platform (BTP), an Ubuntu based OS
  containing features to help develop and deliver hands-on bioinformatics
  training in a more collaborative fashion. In addition, we hope to provide
  a framework through which people can shared and access workshop content as
  well as configure the BTP quickly and easily for such a workshop. This is
  still in development as a proof-of-concept.

Course Overview
===============
A 2-day hands-on Next Generation Sequencing (NGS) Workshop developed by
Bioplatforms Australia (BPA), the Commonwealth Scientific and Industrial
Research Organisation (CSIRO), EMBL-Australia and the European Bioinformatics
Institute.

This course was designed to familiarise participants with NGS data analysis and
provide hands-on computational experience in analysing NGS data using common
analytical workflows for, ChIP-seq, RNASeq data and de novo genome assembly.

Course content
==============
The course covers data analysis of RNA-Seq and ChIP-Seq experiments and de novo
genome assembly of small genomes. Topics will include: data quality, alignment,
data handling and visualisation, region identification, transcriptome analysis
and statistical analysis.

Target Audience
===============
This practical 2 day hands-on course is designed for bench biologists, PhD
students or early career postdoctoral researchers with no or basic bioinformatics
experience who are planning or currently using NGS approaches in their research
area.

Registration/Application
========================
Because places are limited (approx. 30 places), registration is via a selection
process. Therefore applicants should be aware that, in order to be considered
for a place on this course, they MUST complete an application form.

Computer Lab Setup
==================
This workshop has been designed to be run on VM's on Australia's NeCTAR
Research Cloud. Trainees, then use the NoMachine NX Client to provide a remote
dekstop like connection to their own dedicated VM for the duration of the
workshop.

This approach means that the workshop can be delievered in a road-show style
form around Australia, without being limited by local computational resources.
The level of IT support, required of the local hosting institution, is minimal
and only requires the installation of the NoMachine NX Client and SSH firewall
access over port 22 TCP.

We has also developed an Ubuntu 12.04 Image, based on Cloud BioLinux, for the
purpose of running this workshop.